lib/light.js

```diff
@@ -46,9 +48,68 @@ function clampColorTemp(colorTemp, colorTempMin, colorTempMax, logger) {
     return colorTemp;
 }
 
+/**
+ * Converts converter's value input to {x,y}, {h,s,v}, {h,s}, {h} or {s}
+ * @param {object} value converter value parameter
+ * @return {object} normalized color information
+ */
+function extractColorFromValue(value) {
+    if (value.hasOwnProperty('x') && value.hasOwnProperty('y')) {
+        return value;
+    } else if (value.hasOwnProperty('r') && value.hasOwnProperty('g') && value.hasOwnProperty('b')) {
+        return utils.rgbToXY(value.r, value.g, value.b);
+    } else if (value.hasOwnProperty('rgb')) {
+        return utils.rgbToXY(...value.rgb.split(',').map((i) => parseInt(i)));
+    } else if (value.hasOwnProperty('hex') || (typeof value === 'string' && value.startsWith('#'))) {
+        return utils.hexToXY(typeof value === 'string' && value.startsWith('#') ? value : value.hex);
+    } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s') && value.hasOwnProperty('l')) {
+        return utils.hslToHSV(value.h, value.s, value.l);
+    } else if (value.hasOwnProperty('hsl')) {
+        return utils.hslToHSV(...value.hsl.split(',').map((i) => parseInt(i)));
+    } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s') && value.hasOwnProperty('b')) {
+        return {h: value.h, s: value.s, v: value.b};
+    } else if (value.hasOwnProperty('hsb')) {
+        const hsv = value.hsb.split(',').map((i) => parseInt(i));
+        return {h: hsv[0], s: hsv[1], v: hsv[2]};
+    } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s') && value.hasOwnProperty('v')) {
+        return value;
+    } else if (value.hasOwnProperty('hsv')) {
+        const hsv = value.hsv.split(',').map((i) => parseInt(i));
+        return {h: hsv[0], s: hsv[1], v: hsv[2]};
```

Comment:
nit: return these as 'hue', 'saturation', and 'brightness' to stick close to ZCL (if v is 0-100, we can map it to 0-254)

Reply:
technically I think we need to return this as `{"color": {"hue": x, "saturation": x}, "brightness": x}

Reply:
Yeah its a mess, must admit.

Some data format standardization is definitely needed here.

Currently color data is passed in many different formats with different value ranges.

Here I decided to go with `{h, s, v} (0..360, 0..100, 0..100)` because its most common format in `lib/util.js` and its basically target agnostic, representing hue in degrees, saturation and value in percent are universal, hence easy to understand when creating new integrations.

For example Tuya light source API uses string encoded HSV with ranges: `hue (0..360), saturation (0..1000), value (0..1000)` and separate call setting brightness in range `(0..1000)`. Brightness can be set both by `value` component of color and `brightness`.

Conversion to ZCL's `{hue, saturation} (0..254/0..65534, 0..254)` + `brightness (0..255)` is done only in ZCL specific part of the code (i.e. `toZigbee.light_color`.

IMHO its best to keep universal ranges in `lib/` and convert it always from base ranges in device (API) specific code. I can of course change the code to be closer to ZCL and then modify Tuya code to work with it if thats preffered.

Reply:
Maybe I could move `extractColorFromValue` to `lib/utils.js` to avoid confusion that its related only to ZCL.

Reply:
I tend to have the ZCL open when working on things as reference. I’ve had issues a few times were internally name something that is also used in ZCL but in a different content or value range.

Given to the nature of the project I feel that sticking to ZCL defined names/structures would be preferred to avoid confusing and that makes referencing the ZCL easier. E.g we store xy without the Y component because that is what the ZCL does. So storing hue/saturation without either the brightness or level value seems consistent with that. We can store that info in the brightness property which (confusingly sometimes) we map to level later.

That being said, I have no strong preference to that over h,s,v + x,y either. Bu

Reply:
Been thinking about this a bit in my walk.

How about:
- move all color conversion functions to a new lib/color
- Have the extract function spit out 2 object: {x, y} and {h, s, l}, one of them or both can be null when there is no usable date found. (e.g. passing in only color.x and color.y in the input will have only the xy return object set and the hsl one is null)
- Discision making in the light_color converter can then be simplified to:
    - xy != null => prefer xy commands
    - xy == null && hsl != null => prefer enhancedHue commands, or hue commands if enhancedHue not supported
    - hsl != null => additional level command to update the brightness?
  
  All exported function in lib/color should then take either xy, hsl, or mired as input and return one of those as output

Reply:
> move all color conversion functions to a new lib/color

That's a very good idea. All operations within this file should follow common, target agnostic conventions for params and returns.

I'd propose following data types:

```javascript
/**
 * @typedef {object} ColorHSV
 * @property {number} h hue component (0..360)
 * @property {number} s saturation component (0..100)
 * @property {[number]} v value component (0..100)
 */

/**
 * @typedef {object} ColorXY
 * @property {number} x X component (0..1)
 * @property {number} x Y component (0..1)
 */

/**
 * @typedef {object} ColorRGB
 * @property {number} r red component (0..1)
 * @property {number} g green component (0..1)
 * @property {number} b blue component (0..1)
 */

/**
 * @typedef {object} Color
 * @property {?ColorHSV} hsv
 * @property {?ColorXY} xy
 */

/**
 * Takes HSV color, returns HSV color.
 * @param {ColorHSV} color
 * @returns {ColorHSV}
 */
function hsvOperation(color) {
    /* ... */
}

/**
 * Takes HSV color, returns RGB color.
 * @param {ColorHSV} hsv
 * @returns {ColorRGB}
 */
function hsvToRgbOperation(hsv) {
    /* ... */
}

/**
 * Extract Color information from convertes input value.
 * @param {object} value
 * @returns {Color}
 */
function extractColorFromValue(value) {
    /* ... */
}
```

Reply:
Or `hue, saturation, value` for HSV, but then I feel that we should also rename `r, g, b` to `red, green, blue`.

Reply:
SGTM, sorry for increasing the complexity of the PR a bit, but I think overall this should make make things a lot cleaner!

Reply:
np, that's a good review

I already started the rewrite and added progress so far in new commit.

I'm marking this PR as a draft for now.

Reply:
@sjorge there is a problem with "hue" and "saturation" in state. Home Assistant doesn't seem to like it.

When I report "h" and "s" it works fine, but with "hue" and "saturation" color wheel feedback doesn't work.

With "h" and "s" dot is in the right place:

![Zrzut ekranu 2021-04-7 o 09 06 05](https://user-images.githubusercontent.com/13185945/113825039-ca441680-9780-11eb-9630-ff6ac4feb901.png)

With "hue" and "saturation" it doesn't move anywhere:

![Zrzut ekranu 2021-04-7 o 09 07 09](https://user-images.githubusercontent.com/13185945/113824942-a84a9400-9780-11eb-9557-b6d5b536655a.png)

I don't know if its a bug in HA or expected behavior though

Reply:
@Koenkk any idea? Maybe something to do with how the expose data is setup?

Reply:
Expose setup sets "hue" and "saturation" for sure:

https://github.com/Koenkk/zigbee-herdsman-converters/blob/22e7e9316a1356ae255b1688bc1eb8381e331ac5/lib/exposes.js#L288-L290

Reply:
https://github.com/Koenkk/zigbee2mqtt/blob/abd8a0926403d2b6d6c9a1a175f917262926c72f/lib/extension/homeassistant.js#L87-L101

You're not also exposing colorXY right? That seems to take priority if you do.

Reply:
No, I'm definitely not, this is HA config:

```json
{
    "availability": [
        {
            "topic": "zigbee2mqtt/bridge/state"
        }
    ],
    "brightness": true,
    "brightness_scale": 254,
    "color_temp": true,
    "command_topic": "zigbee2mqtt/light_livingroom_ceiling/set",
    "device": {
        "identifiers": [
            "zigbee2mqtt_0x60a423fffe043cc3"
        ],
        "manufacturer": "TuYa",
        "model": "Zigbee & RF 5 in 1 LED Controller WZ5 (RGB+CCT mode) (TS0601_led_5ch_rgbcct)",
        "name": "light_livingroom_ceiling",
        "sw_version": "Zigbee2MQTT 1.18.1"
    },
    "hs": true,
    "json_attributes_topic": "zigbee2mqtt/light_livingroom_ceiling",
    "max_mireds": 454,
    "min_mireds": 250,
    "name": "light_livingroom_ceiling",
    "schema": "json",
    "state_topic": "zigbee2mqtt/light_livingroom_ceiling",
    "unique_id": "0x60a423fffe043cc3_light_zigbee2mqtt",
    "xy": false
}
```

Reply:
Here, HA actually expects "h" and "s":

https://github.com/home-assistant/core/blob/6ec8e17e7b09449cd2a25f80c21732e8753b1e71/homeassistant/components/mqtt/light/schema_json.py#L231-L232

Reply:
Interesting.... https://www.home-assistant.io/integrations/light.mqtt/#hs-color

homeassistant seems to want h & s specifically and there seems to be no way of setting a custom value to grab as you can with some properties.

That probably explains issues of hs bulbs getting out of sync with HA, currently z2m copies the payload to the state.
So if you do a `../set/color` with `{"h": 120, "s": 100, "v": 100}` it would copy these fields ot the state. This would keep HA happy.

Of course when reporting is configured, the fromZigbee converter would get the report and set the `hue` and `saturation` fields for the color object. As long as the color was set by z2m this is not an issue. We just have both `hue` and `h` in the state. (Or when not using HA, most likely just `hue` as that is what I set for my homekit integration).

Same for doing a `.../get/color` as the fromZigbee convert would again set `hue`, `saturation` and/or `x`, and `y`.

But once a color changing remote comes into play (then you have to rely on a read/report) they colors would not be in sync, as the report could change it to hue=0, saturation=100 for red for example, but the h and s values HA is looking for are not updated.

The syncColor PR should fix that though, as it will copy whatever the active colorMode's values are to everything else it finds in the state.

@Koenkk that probably means we should in addition to the state, also check if homeassitant is set to true and always generate h + s values in the color payload if it is. As there could be situations where a bulb supports both and we only ever set XY via z2m but then use a remote that uses HS. In that case colorMode would change and we will sync hue/saturation -> x/y as that is in the state payload. But we won't generate h/s which is needed for HA.

Alternatively we just rename hue -> h and saturation -> s across the board ? although that would break non HA stuff.

Reply:
> Here, HA actually expects "h" and "s":
> 
> https://github.com/home-assistant/core/blob/6ec8e17e7b09449cd2a25f80c21732e8753b1e71/homeassistant/components/mqtt/light/schema_json.py#L249-L250

Yeah looks like we both came to the same conclusion at the same time, that as I mentioned also explained some weirdness when using color controling remotes I've seen reported.

Reply:
BTW I found this: https://github.com/Koenkk/zigbee2mqtt/issues/6983

Reply:
We are already passing the color_mode properly in the replies, z2m should be good on that front

Reply:
> @Koenkk that probably means we should in addition to the state, also check if homeassitant is set to true and always generate h + s values in the color payload if it is. As there could be situations where a bulb supports both and we only ever set XY via z2m but then use a remote that uses HS. In that case colorMode would change and we will sync hue/saturation -> x/y as that is in the state payload. But we won't generate h/s which is needed for HA.
>
> Alternatively we just rename hue -> h and saturation -> s across the board ? although that would break non HA stuff.

Or try to convince HA to support both "h"+"s" and "hue"+"saturation" in their code.

Reply:
That is also an option, but I’m not sure how open they would be to that.

They can’t cater to all possible software out there. The again they shouldn’t expect other software to cater to them either.

Reply:
I think this is out of scope for this PR and it can be picked up with the color sync pr from @sjorge

Reply:
> I think this is out of scope for this PR and it can be picked up with the color sync pr from @sjorge . Let's keep the focus on refactoring and not behavioural changes (as this PR is already quite big)

I agree with this point. In https://github.com/Koenkk/zigbee-herdsman-converters/pull/2453/commits/209d26d7086d1205c41777b2b310b8d349dd137b I changed default behaviour to return {h,s,v} as it was happening previously.